opal/corelib/runtime.js

```diff
@@ -1068,24 +1063,50 @@
 
   Opal.find_obj_super_dispatcher = function(obj, jsid, current_func) {
     var klass = obj.$$meta || obj.$$class;
+
+    // first we need to find the class/module current_func is located on
+    klass = Opal.find_owning_class(klass, current_func);
+
+    if (!klass) {
+      throw new Error("could not find current class for super()");
+    }
+
     jsid = '$' + jsid;
+    return Opal.find_super_func(klass, jsid, current_func);
+  };
+
+  Opal.find_owning_class = function(klass, current_func) {
+    var owner = current_func.$$owner;
 
-    // first we need to find the class/module current_func is defined/donated on
     while (klass) {
-      // when a module has been included in a class, look for that
-      if ((klass.$$iclass && klass.$$module === current_func.$$owner) || klass === current_func.$$owner) {
+      // repeating for readability
+
+      if (klass.$$iclass && klass.$$module === current_func.$$donated) {
+        // this klass was the last one the module donated to
+        // case is also hit with multiple module includes
+        break;
+      }
+      else if (klass.$$iclass && klass.$$module === owner) {
+        // module has donated to other classes but klass isn't one of those
+        break;
+      }
+      else if (owner.$$is_singleton && klass === owner.$$singleton_of.$class()) {
+        // cases like stdlib `Singleton::included` that use a singleton of a singleton
+        break;
```

Comment:
@elia or @meh - Open to suggestions if there is a better way for this singleton case. Only a few weird cases hit this. Another one besides `Singleton` in stdlib was the alias case here: https://github.com/ruby/spec/commit/bbe5d48e6eff38c980010b60b7c42d1acd6316fd

I did at least try and make the boolean flag first to avoid the `$class()` function call unless we need to.